### PR TITLE
PatternToken rules: match only from likely tokens if possible

### PR DESCRIPTION
**Pull request by donnerpeter**
track token indices by lemma/token text in AnalyzedSentence

if a pattern has a token with token/lemma hints and fixed offset from the start,
use these indices to narrow down the search space

7-10% speedup in `Main -l en` for me

### CODE REVIEW

**Comment by donnerpeter** [comment]
@danielnaber ping? :)

**Comment by danielnaber** [comment]
@donnerpeter I'll merge it after release (https://dev.languagetool.org/roadmap), we're in feature-freeze now so that only bug fixes are supposed to be added.

File: languagetool-core/src/main/java/org/languagetool/AnalyzedSentence.java [hunk 1]
```diff
@@ -309,7 +322,29 @@ public String getAnnotations() {
...
    */
   public Set<String> getLemmaSet() {
-    return lemmaSet;
+    return lemmaOffsets.keySet();
+  }
+
+  /**
+   * @return all offsets in {@link #getTokensWithoutWhitespace()} where tokens with the given text occur (case-insensitive),
+   * or {@code null} if there are no such occurrences
+   * @since 5.2
```
Comment (donnerpeter) [inline languagetool-core/src/main/java/org/languagetool/AnalyzedSentence.java:331]: This (and below) probably should be changed to 5.3

Comment (danielnaber) [reply languagetool-core/src/main/java/org/languagetool/AnalyzedSentence.java:331]: Thanks, fixed

**Pull request merged**